common/crypto/platform/openssl/crypto.cpp

```diff
@@ -0,0 +1,165 @@
+// Copyright 2023 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+#include <common/crypto.hpp>
+
+#include <cstdint>
+#include <string>
+#include <vector>
+#include <memory>
+
+#include <common/io.hpp>
+#include <common/error.hpp>
+#include <common/expected.hpp>
+#include <common/common.hpp>
+
+#include <artifact/sha/sha.hpp>
+
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+#include <openssl/err.h>
+
+
+namespace mender {
+namespace common {
+namespace crypto {
+
+const size_t MENDER_DIGEST_SHA256_LENGTH = 32;
+
+using namespace std;
+
+namespace error = mender::common::error;
+namespace io = mender::common::io;
+
+auto pkey_ctx_free_func = [](EVP_PKEY_CTX *ctx) {
+	if (ctx) {
+		EVP_PKEY_CTX_free(ctx);
+	}
+};
+
+auto pkey_free_func = [](EVP_PKEY *key) {
+	if (key) {
+		EVP_PKEY_free(key);
+	}
+};
+auto bio_free_func = [](BIO *bio) {
+	if (bio) {
+		BIO_free(bio);
+	}
+};
+
+auto EncodeBase64(vector<uint8_t> to_encode) -> expected::ExpectedString {
```

Comment:
Is there a reason to use the `auto` for top-level functions? I think we should try to stick to the standard style for best readability.